csrc/multidevice/collective.cpp

```diff
@@ -0,0 +1,199 @@
+    TORCH_INTERNAL_ASSERT(
+        it != params_.team.end(),
+        "root (device ",
+        params_.root,
+        ") must be present in the collective's team");
+    // pytorch's process group expects the root to be specified
+    // by its relative rank, i.e., the device index's offset within the team
+    root_rank_ = std::distance(params_.team.begin(), it);
+  }
+}
+
+std::string Collective::toString(int indent) const {
+  std::stringstream ss;
+  std::string ext_indent(" ", indent);
+  std::string indent1 = ext_indent + "  ";
+  std::string indent2 = ext_indent + "    ";
+
+  ss << ext_indent << "Collective " << collective_type_ << ": {\n";
+
+  if (has_root_) {
+    ss << indent1 << "root: " << params_.root << ",\n";
+  }
+  ss << indent1 << "team: {";
+  for (auto r : params_.team) {
+    ss << r << ", ";
+  }
+  ss << indent1 << "}\n";
+  ss << indent1 << "src_bufs: {";
+  for (auto& t : params_.src_bufs) {
+    ss << "\n" << t;
+  }
+  ss << "\n" << indent1 << "}\n";
+  ss << ext_indent << "}";
+
+  return ss.str();
+}
+
+Broadcast::Broadcast(CollectiveParams params)
+    : Collective(params, "broadcast") {}
+
+c10::intrusive_ptr<c10d::Work> Broadcast::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.src_bufs, 1);
+    assertBufferCount(params_.dst_bufs, 0);
+  } else {
+    assertBufferCount(params_.src_bufs, 0);
+    assertBufferCount(params_.dst_bufs, 1);
+  }
+
+  return comm.getTeam(params_.team)
+      ->broadcast(
+          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
+          {.rootRank = root_rank_});
+}
+
+Gather::Gather(CollectiveParams params) : Collective(params, "gather") {
+  assertBufferCount(params_.src_bufs, 1);
+}
+
+c10::intrusive_ptr<c10d::Work> Gather::post(Communicator& comm) {
+  post_common(*this, comm);
+
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.dst_bufs, params_.team.size());
+    buf_list_ = setBufList(params_.dst_bufs);
+  } else {
+    assertBufferCount(params_.dst_bufs, 0);
+  }
+
+  return comm.getTeam(params_.team)
+      ->gather(buf_list_, params_.src_bufs, {.rootRank = root_rank_});
+}
+
+Allgather::Allgather(CollectiveParams params)
+    : Collective(params, "allgather", false) {
+  assertBufferCount(params_.src_bufs, 1);
+  assertBufferCount(params_.dst_bufs, params_.team.size());
+  buf_list_ = setBufList(params_.dst_bufs);
+}
+
+c10::intrusive_ptr<c10d::Work> Allgather::post(Communicator& comm) {
+  post_common(*this, comm);
+  return comm.getTeam(params_.team)->allgather(buf_list_, params_.src_bufs, {});
+}
+
+Scatter::Scatter(CollectiveParams params) : Collective(params, "scatter") {
+  assertBufferCount(params_.dst_bufs, 1);
+}
+
+c10::intrusive_ptr<c10d::Work> Scatter::post(Communicator& comm) {
+  post_common(*this, comm);
+  if (comm.deviceId() == params_.root) {
+    assertBufferCount(params_.src_bufs, params_.team.size());
```

Comment:
Hmm, can't we just scatter a single buffer like MPI_Scatter?

Reply:
we have to specify several buffers, but they get flattened into one, see https://github.com/NVIDIA/Fuser/pull/797#discussion_r1318679040